data/blog/reverse-engineering-ios-to-fix-sdk-crashes.mdx

```diff
@@ -0,0 +1,286 @@
+
+```
+(lldb) po class_getMethodImplementation(
+    NSClassFromString("UIPrintPanelViewController"),
+    NSSelectorFromString("isKindOfClass:")
+)!
+▿ 0x000000024132b1b8
+  - _rawValue : (Opaque Value)
+```
+
+It points to another address `0x000000024132b1b8`, indicating that `isKindOfClass:` has been overridden in this class, likely to support some internal behavior related to split view handling in the printing UI.
+
+# Reverse-Engineering to Prove a Point
+
+At this point we already had strong indicators that the `UIPrintPanelViewController` implementation of `isKindOfClass:` returns `true` when comparing to a `UISplitViewController`, even though it isn't a subclass of it at all.
+
+But we wanted to be (more) sure, so Itay started to decompile the method in the private framework using [Hopper](https://www.hopperapp.com/) and found a suspicious `usingSplitView` with some register operations in the implementation:
+
+```objc
+/* @class UIPrintPanelViewController */
+-(int)isKindOfClass:(int)arg2 {
+    // Regular registers stuff
+
+    // Probably the super.isKindOfClass
+    r0 = loc_25fdf2cf0(&var_30, 0x1fab10310, arg2);
+
+    r20 = r0;
+
+    // Checks for some special class, possibly UISplitViewController
+    if (loc_25fdf2d00(*0x2782877d0) == arg2) {
+            r20 = [r19 usingSplitView];
+    }
+    r0 = r20;
+    return r0;
+}
+```
+
+Another strong indicator that the implementation of this method is somehow performing special handling for `UISplitViewController` resulting in the invalid cast, crashing SDKs.
+
+This was reason enough to stop the investigation and conclude that this special handling creates an incompatibility with our SDK's assumptions about type casting behavior.
+
+# The Fix
+
+Now that we know what the issue is and that it's pretty much outside of our control what Apple does internally, we resort to the common approach of using [`respondsToSelector:`](<https://developer.apple.com/documentation/objectivec/nsobjectprotocol/responds(to:)>) to [add a runtime check](https://github.com/getsentry/sentry-cocoa/blob/33c961951f46954fae53b1703405746f9836862c/Sources/Swift/Helper/SentryApplicationExtensions.swift#L140-L153) if the instance actually has a method named `viewControllers`:
+
+```swift
+if let splitViewController = vc as? UISplitViewController {
+    // Check if the selector exists as a double-check mechanism
+    // See: https://github.com/getsentry/sentry-cocoa/issues/6725
+    if !splitViewController.responds(to: NSSelectorFromString("viewControllers")) {
+        SentrySDKLog.warning("Failed to get viewControllers from UISplitViewController. This is a known incompatibility in iOS 26.1")
+    } else if splitViewController.viewControllers.count > 0 {
+        return splitViewController.viewControllers
+    }
+}
+```
+
+This might not be the prettiest code, but at least the SDK is not crashing anymore and we were able to quickly release a hotfix to handle this edge case.
```

Comment:
```suggestion
To quote our Director of Engineering [Daniel](https://github.com/HazAT):

> _"We write the dirty code, so our customers don't have to!"_

...and this might not be the prettiest code, but at least the SDK is not crashing anymore and we were able to quickly release a hotfix to handle this edge case
```